### PR TITLE
feat: separate methods for object key value

### PR DESCRIPTION
**Pull request by xDivisionByZerox**
Created in relation to #492.

Thinks I have done:
- Create and implement 2 new methods `faker.random.objectValue` and `faker.random.objectKey`
- Wrote documentation for both methods
- Wrote test for both methods
- Added a deprecation flag and log for `faker.random.objectElement`

### CODE REVIEW

**Review by ST-DDT — APPROVED** [review]

File: src/random.ts [hunk 1]
```diff
@@ -170,6 +170,8 @@ export class Random {
    * faker.random.objectElement(object) // 42
    * faker.random.objectElement(object, 'key') // 'keyB'
    * faker.random.objectElement(object, 'value') // 'valueA'
+   *
+   * @deprecated
```
Comment (ST-DDT) [inline src/random.ts:174]: Please add `@see` references to the new methods.

Comment (Shinigami92) [reply src/random.ts:174]: Do we need to add this annotation and the see to both JSDocs? I think so.

Comment (xDivisionByZerox) [reply src/random.ts:174]: I'm more concerned why there are 2 JSDocs in the first place. Is this some kind of merge problem?

Comment (Shinigami92) [reply src/random.ts:174]: Was added here: https://github.com/faker-js/faker/pull/596
Sadly I think I discussed with @ST-DDT in voice chat without writing a comment about the why
It was something like the one is shown in VSCode and the other is picked up from TypeDoc

Comment (xDivisionByZerox) [reply src/random.ts:174]: Added the `@see` decorator to both JSDocs in 30711140ee4a7380853e7b4c91a0170aacfa694e

File: src/random.ts [hunk 2]
```diff
@@ -189,6 +194,9 @@ export class Random {
    * If this is set to `'value'`, this method will a return a random value of the given instance.
    * Defaults to `'value'`.
    *
+   * @see faker.random.objectKey()
+   * @see faker.random.objectValue()
+   *
```
Comment (ST-DDT) [inline src/random.ts:199]: This one is missing the deprecated flag.

Comment (ST-DDT) [reply src/random.ts:199]: Please also add a `// For typedoc/api docs` comment above this jsdocs.
That way it's easy to tell why it is here.

Comment (xDivisionByZerox) [reply src/random.ts:200]: Should be done now

**Review by ST-DDT — DISMISSED** [review]

**Review by Shinigami92 — DISMISSED** [review]

File: src/random.ts [hunk 3]
```diff
@@ -203,10 +214,48 @@ export class Random {
     object = { foo: 'bar', too: 'car' } as unknown as T,
     field = 'value'
   ): K | T[K] {
+    console.warn(
+      'Deprecation Warning: faker.random.objectElement is now split into 2 seperate methods faker.datatype.objectKey and faker.datatype.objectValue'
```
Comment (import-brain) [inline src/random.ts:218]:
```suggestion
      'Deprecation Warning: faker.random.objectElement() is now split into 2 separate methods: faker.datatype.objectKey() and faker.datatype.objectValue().'
```

Comment (import-brain) [reply src/random.ts:218]: Very small change, but I think this makes the warning slightly easier to read.

**Comment by xDivisionByZerox** [comment]
The failing CI doesn't seem to be related to the changes I made, rather than with the `fake` module.
```
2022-03-26T09:57:08.4524974Z  > Fake.fake src/fake.ts:85:12
2022-03-26T09:57:08.4525598Z
2022-03-26T09:57:08.4526162Z     if (this.faker[parts[0]] == null) {
2022-03-26T09:57:08.4527269Z           throw new Error('Invalid module: ' + parts[0]);
2022-03-26T09:57:08.4527942Z                             ^
2022-03-26T09:57:08.4528536Z     }

Error: invalid module: {datatype
```

I guess a rebase could fix this. (?)

**Review by Shinigami92 — CHANGES_REQUESTED** [review]
1. We need to add test for the deprecation warning
2. We need to swap out all internal usages like in `finance.currencyCode` (please use search to find more occurrences)

**Comment by Shinigami92** [comment]
> The failing CI doesn't seem to be related to the changes I made, rather than with the `fake` module.
> 
> ```
> 2022-03-26T09:57:08.4524974Z  > Fake.fake src/fake.ts:85:12
> 2022-03-26T09:57:08.4525598Z
> 2022-03-26T09:57:08.4526162Z     if (this.faker[parts[0]] == null) {
> 2022-03-26T09:57:08.4527269Z           throw new Error('Invalid module: ' + parts[0]);
> 2022-03-26T09:57:08.4527942Z                             ^
> 2022-03-26T09:57:08.4528536Z     }
> 
> Error: invalid module: {datatype
> ```
> 
> I guess a rebase could fix this. (?)

I'm not totally sure what's going on here, I hope @ST-DDT has some more in-depth inside into the fake function and what could be the problem here related to this PR changes
`main` branch is green and I did not see this previously somewhere

Here is a local failed test result

```js
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  test/all_functional.spec.ts > faker.fake functional tests > uk > datatype > array()
Error: Invalid module: {datatype
 ❯ Fake.fake src/fake.ts:85:12
     83| 
     84|     if (this.faker[parts[0]] == null) {
     85|       throw new Error('Invalid module: ' + parts[0]);
       |            ^
     86|     }
     87| 
 ❯ Fake.fake src/fake.ts:124:16
 ❯ test/all_functional.spec.ts:96:35

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯
```

**Comment by ST-DDT** [comment]
I have identified the issue.
The generated string contains a `$` which causes the string replace method to insert the token again. In this case the `$` was directly after a `{` which then caused the issue.
I will provide a fix for the faker method.

**Comment by ST-DDT** [comment]
Fix and test provided: https://github.com/faker-js/faker/pull/688/files#diff-d764addde9b944dba41e8666051bf020b5f44349c6043d6d7770bd58073eca25R83

File: src/random.ts [hunk 4]
```diff
@@ -203,10 +214,48 @@ export class Random {
     object = { foo: 'bar', too: 'car' } as unknown as T,
     field = 'value'
   ): K | T[K] {
+    console.warn(
+      'Deprecation Warning: faker.random.objectElement() is now split into 2 separate methods: faker.datatype.objectKey() and faker.datatype.objectValue().'
+    );
```
Comment (ST-DDT) [inline src/random.ts:219]: This should use the new deprecation helper method.

**Review by ST-DDT — CHANGES_REQUESTED** [review]

File: src/random.ts (hunk 4 above)
Comment (ST-DDT) [reply src/random.ts:219]: @xDivisionByZerox This is almost ready to merge. Do you have time to address this soonish or should we change it for you?

**Comment by Shinigami92** [comment]
I just found out that we need JSDoc with `@deprecated` on ALL signatures!



In the one line with not using 'key', it is not strikedthrough...

File: src/random.ts [hunk 5]
```diff
@@ -194,24 +212,67 @@ export class Random {
...
+   * faker.random.objectKey({ myProperty: 'myValue' }) // 'myProperty'
+   */
+  objectKey<T extends Record<string, unknown>>(
+    object: T = { foo: 'bar', too: 'car' } as unknown as T
+  ): keyof T {
     const array: Array<keyof T> = Object.keys(object);
-    const key = this.arrayElement(array);
+    const key = this.faker.random.arrayElement(array);
+
+    return key;
```
Comment (ST-DDT) [inline src/random.ts:281]: IMO you can directly return here instead of storing it in const.

File: src/random.ts [hunk 6]
```diff
@@ -194,24 +212,67 @@ export class Random {
...
+   * faker.random.objectValue({ myProperty: 'myValue' }) // 'myValue'
+   */
+  objectValue<T extends Record<string, unknown>>(
+    object: T = { foo: 'bar', too: 'car' } as unknown as T
+  ): T[keyof T] {
+    const key = this.faker.random.objectKey(object);
+    const value = object[key];
 
-    return field === 'key' ? (key as K) : (object[key] as T[K]);
+    return value;
```
Comment (ST-DDT) [inline src/random.ts:299]: Same here.

File: src/random.ts [hunk 7]
```diff
@@ -194,24 +211,92 @@ export class Random {
...
+   * Returns a random key from given object.
+   *
+   * @param object The optional object to be used. If not specified, it defaults to `{ foo: 'bar', too: 'car' }`.
+   *
+   * @example
+   * faker.random.objectKey() // 'foo'
+   * faker.random.objectKey({ myProperty: 'myValue' }) // 'myProperty'
+   */
+  objectKey<T extends Record<string, unknown>>(
+    object: T = { foo: 'bar', too: 'car' } as unknown as T
```
Comment (ST-DDT) [inline src/random.ts:276]: IMO this method should not have a default value.

File: src/random.ts [hunk 8]
```diff
@@ -194,24 +211,92 @@ export class Random {
...
+   * Returns a random value from given object.
+   *
+   * @param object The optional object to be used. If not specified, it defaults to `{ foo: 'bar', too: 'car' }`.
+   *
+   * @example
+   * faker.random.objectValue() // 'bar'
+   * faker.random.objectValue({ myProperty: 'myValue' }) // 'myValue'
+   */
+  objectValue<T extends Record<string, unknown>>(
+    object: T = { foo: 'bar', too: 'car' } as unknown as T
```
Comment (ST-DDT) [inline src/random.ts:294]: IMO this method should not have a default value.

**Review by ST-DDT — CHANGES_REQUESTED** [review]

**Review by ST-DDT — APPROVED** [review]

**Review by Shinigami92 — APPROVED** [review]

**Pull request merged**